From a336a9d7a7b8961274689441f981ed863ddb5758 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 20 Feb 2024 11:28:10 +0100
Subject: [PATCH 0619/1695] drivers: stm32_tamp: support STM32MP21x SoC family

Add support for STM32MP21x SoC family.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0d1d361d71ef18cd48f56ff98118fccc8a6f069e drivers: stm32_tamp: support STM32MP21x SoC family]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I750833dc0eb577822a08d3f5757fbba788a7b084
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361569
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_tamp.c | 70 ++++++++++++++++++++++++++++++++++++++-
 1 file changed, 69 insertions(+), 1 deletion(-)

diff --git a/core/drivers/stm32_tamp.c b/core/drivers/stm32_tamp.c
index ca54ccdf9..9a51af6dd 100644
--- a/core/drivers/stm32_tamp.c
+++ b/core/drivers/stm32_tamp.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright (c) 2021-2022, STMicroelectronics
+ * Copyright (c) 2021-2024, STMicroelectronics
  */
 
 #include <assert.h>
@@ -186,6 +186,9 @@
 #define _TAMP_STM32MP25_OR_IN5RMP_PF6		U(0)
 #define _TAMP_STM32MP25_OR_IN5RMP_PZ4		BIT(2)
 
+#define _TAMP_STM32MP21_OR_IN1RMP_PC4		U(0)
+#define _TAMP_STM32MP21_OR_IN1RMP_PI8		BIT(0)
+
 /* _TAMP_ERCFGR bit fields */
 #define _TAMP_ERCFGR_ERCFG0		BIT(0)
 #define _TAMP_ERCFGR_ERCFG_MASK		BIT(0)
@@ -321,6 +324,27 @@ static const char * const itamper_name[] = {
 };
 #endif
 
+#if defined(CFG_STM32MP21)
+static const char * const itamper_name[] = {
+	[INT_TAMP1] = "Backup domain voltage threshold monitoring",
+	[INT_TAMP2] = "Temperature monitoring",
+	[INT_TAMP3] = "LSE monitoring",
+	[INT_TAMP4] = "HSE monitoring",
+	[INT_TAMP5] = "RTC Calendar overflow",
+	[INT_TAMP6] = "JTAG TAP access in secured-closed",
+	[INT_TAMP7] = "ADC2 analog watchdog monitoring1",
+	[INT_TAMP8] = "Monotonic counter 1 overflow",
+	[INT_TAMP9] = "Cryptographic peripherals fault",
+	[INT_TAMP10] = "Monotonic counter 2 overflow",
+	[INT_TAMP11] = "IWDG3 reset",
+	[INT_TAMP12] = "ADC2 analog watchdog monitoring2",
+	[INT_TAMP13] = "ADC2 analog watchdog monitoring3",
+	[INT_TAMP14] = "RIFSC or BSEC or DBGMCU fault",
+	[INT_TAMP15] = "IWDG1_reset",
+	[INT_TAMP16] = "BOOTROM fault",
+};
+#endif
+
 static struct stm32_tamp_conf int_tamp_mp13[] = {
 	{ .id = INT_TAMP1 }, { .id = INT_TAMP2 }, { .id = INT_TAMP3 },
 	{ .id = INT_TAMP4 }, { .id = INT_TAMP5 }, { .id = INT_TAMP6 },
@@ -343,6 +367,15 @@ static struct stm32_tamp_conf int_tamp_mp25[] = {
 	{ .id = INT_TAMP15 },
 };
 
+static struct stm32_tamp_conf int_tamp_mp21[] = {
+	{ .id = INT_TAMP1 }, { .id = INT_TAMP2 }, { .id = INT_TAMP3 },
+	{ .id = INT_TAMP4 }, { .id = INT_TAMP5 }, { .id = INT_TAMP6 },
+	{ .id = INT_TAMP7 }, { .id = INT_TAMP8 }, { .id = INT_TAMP9 },
+	{ .id = INT_TAMP10 }, { .id = INT_TAMP11 }, { .id = INT_TAMP12 },
+	{ .id = INT_TAMP13 }, { .id = INT_TAMP14 }, { .id = INT_TAMP15 },
+	{ .id = INT_TAMP16 },
+};
+
 static struct stm32_tamp_conf ext_tamp_mp13[] = {
 	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
 	{ .id = EXT_TAMP4 }, { .id = EXT_TAMP5 }, { .id = EXT_TAMP6 },
@@ -365,6 +398,12 @@ static struct stm32_tamp_conf ext_tamp_mp23[] = {
 	{ .id = EXT_TAMP7 },
 };
 
+static struct stm32_tamp_conf ext_tamp_mp21[] = {
+	{ .id = EXT_TAMP1 }, { .id = EXT_TAMP2 }, { .id = EXT_TAMP3 },
+	{ .id = EXT_TAMP4 }, { .id = EXT_TAMP5 }, { .id = EXT_TAMP6 },
+	{ .id = EXT_TAMP7 },
+};
+
 #define GPIO_BANK(port)	 ((port) - 'A')
 
 static const struct stm32_tamp_pin_map pin_map_mp13[] = {
@@ -456,6 +495,17 @@ static const struct stm32_tamp_pin_map pin_map_mp25[] = {
 	},
 };
 
+static const struct stm32_tamp_pin_map pin_map_mp21[] = {
+	{
+		.id = EXT_TAMP1, .bank = GPIO_BANK('I'), .pin = 8,
+		.out = false, .conf = _TAMP_STM32MP21_OR_IN1RMP_PI8,
+	},
+	{
+		.id = EXT_TAMP1, .bank = GPIO_BANK('C'), .pin = 4,
+		.out = false, .conf = _TAMP_STM32MP21_OR_IN1RMP_PC4,
+	},
+};
+
 /* Expects at most a single instance */
 static struct stm32_tamp_instance stm32_tamp;
 
@@ -2149,9 +2199,27 @@ static const struct stm32_tamp_compat mp23_compat = {
 		.pin_map_size = ARRAY_SIZE(pin_map_mp25),
 };
 
+static const struct stm32_tamp_compat mp21_compat = {
+		.nb_monotonic_counter = 2,
+		.tags = TAMP_HAS_REGISTER_SECCFGR |
+			TAMP_HAS_REGISTER_PRIVCFGR |
+			TAMP_HAS_RIF_SUPPORT |
+			TAMP_HAS_REGISTER_ERCFGR |
+			TAMP_HAS_REGISTER_CR3 |
+			TAMP_HAS_REGISTER_ATCR2 |
+			TAMP_HAS_CR2_SECRET_STATUS,
+		.int_tamp = int_tamp_mp21,
+		.int_tamp_size = ARRAY_SIZE(int_tamp_mp21),
+		.ext_tamp = ext_tamp_mp21,
+		.ext_tamp_size = ARRAY_SIZE(ext_tamp_mp21),
+		.pin_map = pin_map_mp21,
+		.pin_map_size = ARRAY_SIZE(pin_map_mp21),
+};
+
 static const struct dt_device_match stm32_tamp_match_table[] = {
 	{ .compatible = "st,stm32mp25-tamp", .compat_data = &mp25_compat },
 	{ .compatible = "st,stm32mp23-tamp", .compat_data = &mp23_compat },
+	{ .compatible = "st,stm32mp21-tamp", .compat_data = &mp21_compat },
 	{ .compatible = "st,stm32mp13-tamp", .compat_data = &mp13_compat },
 	{ .compatible = "st,stm32-tamp", .compat_data = &mp15_compat },
 	{ }
-- 
2.47.2

